### PR TITLE
Product Preview: Add WebKitViewController to handle authenticated sessions

### LINKED ISSUE #7911: Product Preview: Webview Authentication (WP.com hosted sites)
To access product preview URL we need to authenticate current user in the context of WebView.
The first milestone/scope - WP.com based stores (not self-hosted).

Internal ref: pe5pgL-136-p2#comment-1075

- [x] add `loginURL` to `Site` https://github.com/woocommerce/woocommerce-ios/pull/7943
- [x] migrate `CookieJar` from WP (no changes) https://github.com/woocommerce/woocommerce-ios/pull/7961
- [x] migrate `AuthenticationService` from WP (no changes) https://github.com/woocommerce/woocommerce-ios/pull/7961
- [x] migrate `RequestAuthenticator` from WP (tweak for WC objects, trim unnecessary flows) https://github.com/woocommerce/woocommerce-ios/pull/7961
- [ ] migrate `WebKitViewController` that will use updated `RequestAuthenticator` (trim it down to minimal UI and functionality) https://github.com/woocommerce/woocommerce-ios/pull/7974

### PR DESCRIPTION
**Pull request by ealeksandrov**
Closes https://github.com/woocommerce/woocommerce-ios/issues/7911, https://github.com/woocommerce/woocommerce-ios/issues/7946.

## Description

This PR adds `WebKitViewController` that uses access token for current user to authenticate them in the context of `WKWebView`.
Most of the code copied from [WordPress-iOS](https://github.com/wordpress-mobile/WordPress-iOS/).

## Testing

Test on the device, as iOS Simulator has [a bug around cookies management](https://github.com/woocommerce/woocommerce-ios/pull/7974#issuecomment-1300776912).

1. Switch to WP.com store.
2. Open products tab.
3. Start the flow to create new product ("+" button in navbar on product list)
4. Edit the title (or anything else) to have non-blank product.
5. Tap the "Preview" button appeared in navbar.
6. Confirm that webpage for draft product loads correctly.
7. Open user profile and tap "Log out". Close the webview.
8. Tap the "Preview" button in navbar again.
9. Confirm that webpage for draft product again loads correctly, authenticated for current user.

## Video

https://user-images.githubusercontent.com/3132438/199501283-90940aa9-0476-400b-90f7-2b63dafbea91.mp4

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/Authentication/WebAuth/WebKitViewController.swift [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+    @objc lazy var refreshButton: UIBarButtonItem = {
+        let button = UIBarButtonItem(image: .gridicon(.refresh), style: .plain, target: self, action: #selector(WebKitViewController.refresh))
+        button.title = NSLocalizedString("Refresh", comment: "Button label to refresh a web page")
+        return button
+    }()
+    @objc lazy var closeButton: UIBarButtonItem = {
+        let button = UIBarButtonItem(image: .gridicon(.cross), style: .plain, target: self, action: #selector(WebKitViewController.close))
+        button.title = NSLocalizedString("webKit.button.dismiss", value: "Dismiss", comment: "Verb. Dismiss the web view screen.")
+        return button
+    }()
```
Comment (ealeksandrov) [inline WooCommerce/Classes/Authentication/WebAuth/WebKitViewController.swift:77]: While we don't need toolbar on the WebView yet (it's disabled by `secureInteraction` option), I decided to keep the (almost) full functionality. We can improve the preview feature by adding "device" button, already supported on WP-iOS with [PreviewWebKitViewController](https://github.com/wordpress-mobile/WordPress-iOS/blob/383304dc332f1b42db01b95c823b4a0392170907/WordPress/Classes/ViewRelated/Post/Preview/PreviewWebKitViewController.swift) subclass:

mobile|desktop
--|--
Simulator|Simulator

**Comment by rachelmcr** [comment]
I tested this on my [WP.com test store](https://teststore.rachel-test-domain.com/) but unfortunately the preview isn't working for me there. I'm getting redirected to the WP.com login screen:



Should this be working for all WP.com-hosted sites, including sites with custom domains? (The domain is the big difference I see with my test store.)

**Comment by ealeksandrov** [comment]
@rachelmcr can you try the flow on the device? I have the same issue with simulator, but never on the device. Sorry I forgot to mention it in the description.

For the reference, I've traced the issue to these lines:

https://github.com/woocommerce/woocommerce-ios/blob/5e67a0c34c1b421db427cae42cc96efdd9c8dc9d/WooCommerce/Classes/Authentication/WebAuth/AuthenticationService.swift#L188-L189

On the device this code correctly combines response cookies from 2 requests.
On simulator it shows second request cookies in both objects: 
- previous requests cookie storage: `session.configuration.httpCookieStorage?.cookies` (auth cookie is here)
- extraction of the cookies from the latest request: `self.cookies(from: response, ...)`

From the proxy log there is no difference in networking requests/responses.

**Comment by rachelmcr** [comment]
Aha, thanks! I ran into an odd issue on my device at first, where the `Site.loginURL` was an empty string and that resulted in the web view being blank:



(I lost track of the specific console error when that happened, but it was a problem with the web view trying to load the request `?redirect_to=https://teststore.rachel-test-domain.com/?post_type%3Dproduct%26p%3D3955%26preview%3Dtrue`.)

But when I relaunched the app that seemed to correct itself, and now the previews are working consistently. Not sure if that's something worth digging into further, but thought I'd mention it.

I'll finish reviewing the code now!

**Review by rachelmcr — APPROVED** [review]
After I resolved the issues mentioned above, everything tested well here! 👍 

Given the amount of code here that's the same between WPiOS and WCiOS, do you think it's worth refactoring it into a shared repo (maybe the `WordPressUI` pod)? (Not necessarily right now, but I'm thinking about future maintenance.)

**Comment by ealeksandrov** [comment]
> odd issue on my device at first, where the Site.loginURL was an empty string

It happens when the old `Site` object is loaded from Core Data, where the `loginURL` is missing:
https://github.com/woocommerce/woocommerce-ios/blob/47784c564572373d30f672ad7ec6cebe0e6c6ee7/Yosemite/Yosemite/Model/Storage/Site%2BReadOnlyConvertible.swift#L36

Do you think hiding "preview" button for empty `loginURL` is reasonable? After minor migration period (1 app restart?) it will be filled in for everyone. Another approach is to force an API call to refresh the data, but I don't like that.

> Given the amount of code here that's the same between WPiOS and WCiOS, do you think it's worth refactoring it into a shared repo (maybe the WordPressUI pod)? (Not necessarily right now, but I'm thinking about future maintenance.)

Agree! I considered moving this to WPAuthenticator library (also [WP app comment](https://github.com/wordpress-mobile/WordPress-iOS/blob/8c6c6a05b83c9949fd14f7b2aab422c07f56b4eb/WordPress/Classes/Utility/Networking/RequestAuthenticator.swift#L7-L9)), but it will require refactoring of both client apps and migration of additional dependencies. We can do it after we finish and validate this feature. Maybe we'll reuse it in other parts of the app and will have additional requirements by that time.

**Comment by rachelmcr** [comment]
> Do you think hiding "preview" button for empty loginURL is reasonable? After minor migration period (1 app restart?) it will be filled in for everyone. Another approach is to force an API call to refresh the data, but I don't like that.

Agreed, I think just checking for an empty `loginURL` makes sense. Forcing a refresh seems like overkill, especially since it should be resolved fairly quickly.

> I considered moving this to WPAuthenticator library (also [WP app comment](https://github.com/wordpress-mobile/WordPress-iOS/blob/8c6c6a05b83c9949fd14f7b2aab422c07f56b4eb/WordPress/Classes/Utility/Networking/RequestAuthenticator.swift#L7-L9)), but it will require refactoring of both client apps and migration of additional dependencies. We can do it after we finish and validate this feature.

Sounds good! I missed that WP comment before but it sums up the effort pretty well. :)

**Comment by ealeksandrov** [comment]
Added a check in af2e711 + unit test in 43e4dca.

**Pull request merged**